### PR TITLE
Consistent widening in access expressions

### LINKED ISSUE #31728: Strange error involving evolving array type and Array#sort callback
**TypeScript Version:**  3.6.0-dev.20190602 (seems to work with 20190601)

**Search Terms:**

**Code**

```ts
// @noImplicitAny: true
interface I {
    a: string;
    b: string;
}

declare function compare(a: I, b: I): number;

function test() {
    const result = [];
    result.push({foo: true, a: 'a', b: 'b'});
    //           ~~~~~~~~~ error here
    result.sort(compare); // commenting out this line makes the error disappear
}
```

**Expected behavior:**

No error

**Actual behavior:**

```
Argument of type '(a: I, b: I) => number' is not assignable to parameter of type '(a: { foo: boolean; a: string; b: string; }, b: { foo: boolean; a: string; b: string; }) => number'.
  Types of parameters 'a' and 'a' are incompatible.
    Type '{ foo: boolean; a: string; b: string; }' is not assignable to type 'I'.
      Object literal may only specify known properties, and 'foo' does not exist in type 'I'.
```

Maybe the error makes sense with `strictNullChecks` enabled (which is off in the above example), but then the location is completely off

**Playground Link:** 

**Related Issues:**
The only two changes in that version are #31711 and #31687. /cc @ahejlsberg

### LINKED ISSUE #31762: #31711 breaks assignability
```ts
// @strict: true
function g(headerNames: any) {
    let t = [{ hasLineBreak: false, cells: [] }];
    const table = [{cells: headerNames }].concat(t);
}
```

**Expected behavior:**
No error.

**Actual behavior:**
Error on `concat(t)`: `{ hasLineBreak: boolean, cells: never[] }` is not assignable to `{ cells: any }`. `hasLineBreak` is missing.

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR, in a property access `obj.x` or element access `obj["x"]` we widen `obj` when the property or element access is (a) the target of an assignment, or (b) a method access in a call expression (i.e. the this of the call). These are the situations in which we are mutating (or possibly mutating) the object and therefore should widen its type.

Previously we'd always widen `obj` in a property access and never widen `obj` in an element access, which was inconsistent.

Fixes #31728.
Fixes #31762.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot run dt
@typescript-bot test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20217,19 +20217,25 @@ namespace ts {
             return checkPropertyAccessExpressionOrQualifiedName(node, node.left, node.right);
         }
 
+        function isMethodAccessForCall(node: Node) {
+            while (node.parent.kind === SyntaxKind.ParenthesizedExpression) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: This should probably? also skip both kinds of casts.

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
@typescript-bot test this

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: No, if there is an intervening type assertion then that specifies the type and widening isn't relevant.

**Pull request merged**